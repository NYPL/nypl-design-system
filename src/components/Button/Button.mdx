import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";

import * as ButtonStories from "./Button.stories";
import Link from "../Link/Link";

<Meta of={ButtonStories} />

# Button

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `0.0.4`      |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#button-groups" target="_self">Button Groups</Link>}
- {<Link href="#with-icons" target="_self">With Icons</Link>}
- {<Link href="#patterns" target="_self">Patterns</Link>}
- {<Link href="#other-examples" target="_self">Other Examples</Link>}
- {<Link href="#sizes" target="_self">Sizes</Link>}

## Overview

<Description of={ButtonStories} />

Update the `buttonType` in the Controls section to see the types of buttons
that can be rendered.

- `"primary"` is used for actions that move the user forward. This is the default.
- `"secondary"` is used for actions that move the user back, such as cancellations.
- `"text"` is used for equally weighted actions in a text based list.
- `"callout"` is used for call to action text such as "Donate".
- `"noBrand"` is a variant used when there is no brand and will display the
  background color as black.
- `"pill"` is used for equally weighted actions within cards and always in a set.
- `"link"` has been deprecated and should not be used.

When one and only one `Icon` component is passed inside a `Button` component with
no text, it will automatically be configured to use the `"iconOnly"` type.

## Component Props

<Canvas of={ButtonStories.WithControls} />

<ArgTypes of={ButtonStories.WithControls} />

## Accessibility

The Reservoir `Button` component renders an HTML `<button>` element that is
accessible. `Button`s should be used to perform user actions or submit forms,
never to link and go to another page. For `Button`s without text and only an
icon, make sure to pass an `aria-label` value to the `Button`. `Button`s cannot
have images of text.

`Button`s can be rendered in three sizes through the `size` prop: `"small"`,
`"medium"`, and `"large"`. The default size is `"medium"`. Even though the `"small"`
size is available, we recommend to use the `"medium"` and `"large"` sizes.

The `Button` component also has a `screenreaderOnlyText` prop that can be used to
add additional text to the component that is only visible to screen readers. This
can be used to provide additional context to the `Button` when the text is short,
such as visible "Read more..." text.

Resources:

- [W3C WAI ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/#button)
- [W3C WAI ARIA Button Examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html)
- [MDN ARIA: button role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role)
- [Chakra UI Button](https://chakra-ui.com/docs/components/form/button)
- [W3C WAI WCAG Target Size](https://www.w3.org/WAI/WCAG21/Understanding/target-size.html)

## Button Groups

Primary and secondary buttons should be grouped like in the following example.
The `ButtonGroup` component should be used to wrap the `Button` component.

<Canvas of={ButtonStories.ButtonGroups} />

## With Icons

The `Icon` component collaborates great with the `Button` component. Check out
[`src/components/Icons/Icon.tsx`](https://github.com/NYPL/nypl-design-system/blob/development/src/components/Icons/Icon.tsx)
to see the full list of icon names that can be rendered.

The icon can be placed to the left or the right of the button text.

<Canvas of={ButtonStories.TextWithIcon} />

Text in a button is optional if an icon is rendered, but make sure that there
is an `aria-label` passed to the `Button` component `aria-label="Previous"`.

<Canvas of={ButtonStories.IconOnly} />

## Patterns

Buttons are blue by default but should be red when it is intended to be used as
a call out. The "Back to Top" button must be of `buttonType="secondary"` and
must include an up arrow icon on the right side.

<Canvas of={ButtonStories.Patterns} />

## Other Examples

The variations modified by the `buttonType` prop:

<Canvas of={ButtonStories.ButtonTypes} />

Modifying the `isDisabled` prop:

<Canvas of={ButtonStories.DisabledButtons} />

## Sizes

Update the size of the `Button` component with the `size` prop and values:
`"small"`, `"medium"`, `"large"`. The default size is `"medium"`.

<Canvas of={ButtonStories.ButtonSizes} />
