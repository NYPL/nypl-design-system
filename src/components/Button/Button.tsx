import * as React from "react";
import bem from "../../utils/bem";
import { ButtonTypes } from "./ButtonTypes";

interface ButtonProps {
  /** Additional attributes passed to the button */
  attributes?: { [key: string]: any };
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** The kind of button assigned through the `ButtonTypes` enum  */
  buttonType?: ButtonTypes;
  /** Additional className for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  className?: string;
  /** Adds 'disabled' property to the button */
  disabled?: boolean;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** Trigger the Button's action through the `mouseDown` event handler instead of `onClick`. `false` by default. */
  mouseDown?: boolean;
  /** The action to perform on the `<button>`'s onClick function */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent) => void;
  /** The html button attribute */
  type?: "submit" | "button" | "reset";
}

/** Renders a simple `button` element with custom classes and modifiers. */
export default class Button extends React.Component<ButtonProps, any> {
  static defaultProps = {
    mouseDown: false,
  };

  constructor(props: ButtonProps) {
    super(props);
  }

  render(): JSX.Element {
    const {
      attributes,
      blockName,
      buttonType,
      children,
      className,
      disabled,
      id,
      modifiers = [],
      mouseDown,
      onClick,
      type = "submit",
    } = this.props;

    const buttonModifiers = modifiers ? modifiers : [];
    if (!buttonModifiers.find(modifier => modifier === buttonType)) {
      buttonModifiers.push(buttonType);
    }

    const baseClass = "button";

    const btnCallback = mouseDown
      ? { onMouseDown: onClick }
      : { onClick: onClick };

    return (
      <button
        id={id}
        className={`${baseClass} ${bem(baseClass, buttonModifiers, blockName, [
          className,
        ])}`}
        type={type}
        disabled={disabled}
        {...attributes}
        {...btnCallback}
      >
        {children}
      </button>
    );
  }
}
