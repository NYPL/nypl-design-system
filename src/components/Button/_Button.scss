@mixin button-base {
  @include space-inset-xs;

  border-radius: 2px;
  cursor: pointer;
  display: flex;
  justify-content: center;
  text-decoration: none;

  .icon {
    @include icon;

    fill: currentColor;

    &--icon-left {
      margin-right: var(--space-xs);
    }

    &--icon-right {
      margin-left: var(--space-xs);
    }
  }
}

@mixin button--filled {
  @include button-base;

  background-color: var(--ui-link-primary);
  border: none;
  color: var(--ui-white);

  &:visited {
    color: var(--ui-white);
  }

  &:hover {
    background-color: var(--ui-link-secondary);
  }

  &:disabled {
    background-color: var(--ui-gray-light);
    color: var(--ui-gray-dark);
    pointer-events: none;
  }
}

@mixin button--outline {
  @include button-base;

  background-color: var(--ui-white);
  border: 1px solid var(--ui-gray-light);
  color: inherit;

  &:visited {
    color: inherit;
  }

  &:hover {
    background-color: var(--ui-gray-xxlight);
  }

  &:disabled {
    background-color: var(--ui-gray-light);
    color: var(--ui-gray-dark);
    pointer-events: none;
  }
}

@mixin button--link {
  background-color: transparent;
  color: var(--ui-link-primary);
  text-decoration: underline;

  &--no-underline {
    text-decoration: none;
  }

  &:hover {
    background-color: transparent;
    color: var(--ui-link-secondary);
  }

  &:disabled {
    background-color: transparent;
    color: var(--ui-gray-dark);
    pointer-events: none;
  }
}

@mixin button--pill {
  @include button-base;
  @include button--outline;

  border-radius: 20px;
}

.button {
  @include button--filled;

  &--link {
    @include button--link;
  }

  &--outline {
    @include button--outline;
  }

  &--pill {
    @include button--pill;
  }
}
