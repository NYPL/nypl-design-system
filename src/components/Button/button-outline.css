@import "../Icons/_Icons.scss";

@mixin button-base {
  border-radius: 2px;
  cursor: pointer;
  display: flex;
  justify-content: center;
  padding: var(--space-xs);
  text-decoration: none;

  .icon {
    @include icon;

    fill: currentColor;

    &--icon-left {
      margin-right: var(--space-xs);
    }

    &--icon-right {
      margin-left: var(--space-xs);
    }
  }
}

@mixin button--outline {
  @include button-base;

  background-color: var(--ui-white);
  border: 1px solid var(--ui-gray-light);
  color: inherit;

  &:visited {
    color: inherit;
  }

  &:hover {
    background-color: var(--ui-gray-xxlight);
  }

  &:disabled {
    background-color: var(--ui-gray-light);
    color: var(--ui-gray-dark);
  }
}
