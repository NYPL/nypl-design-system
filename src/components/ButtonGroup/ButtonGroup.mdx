import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";

import * as ButtonGroupStories from "./ButtonGroup.stories";

<Meta of={ButtonGroupStories} />

# ButtonGroup

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.28.0`   |
| Latest            | `1.5.0`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Button Width](#button-width)
- [Layout](#layout)
- [isDisabled](#isDisabled)

## Overview

<Description of={ButtonGroupStories} />

Note: The `ButtonGroup` component does not come from Chakra because we want to
display the `Button`s in a column layout as an option, as well as control the
width.

## Component Props

<Canvas of={ButtonGroupStories.WithControls} />

<ArgTypes of={ButtonGroupStories.WithControls} />

## Accessibility

This component is a wrapper around the `Button` component. For accessibility
information on the `Button` component, see the [Button
Accessibility](https://nypl.github.io/nypl-design-system/reservoir/v1/?path=/docs/components-form-elements-button--button-with-controls#accessibility)
documentation.

## Button Width

The `buttonWidth` prop can be used to set the width of the internal `Button`
components. The default value is `default` and it sets the width to
`"fit-content"` so the buttons will expand to its content. The `"full"` value
will set the width to the parent's width.

<Canvas of={ButtonGroupStories.Width} />

## Layout

The `layout` prop can be used to set the layout to either `row` or `column`.

<Canvas of={ButtonGroupStories.Layout} />

## Disabled

The `isDisabled` prop can be used to disable all the `Button` children at once.
Individual `Button`s can still be disabled with their own `isDisabled` prop.

<Canvas name="isDisabled" of={ButtonGroupStories.Disabled} />
