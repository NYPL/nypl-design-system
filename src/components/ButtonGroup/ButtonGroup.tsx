import { chakra, Stack, useStyleConfig } from "@chakra-ui/react";
import React, { forwardRef } from "react";

import Button from "../Button/Button";
import { LayoutTypes } from "../../helpers/types";
import useNYPLBreakpoints from "../../hooks/useNYPLBreakpoints";

export const buttonGroupWidthsArray = ["default", "full"] as const;
export type ButtonGroupWidths = typeof buttonGroupWidthsArray[number];

interface ButtonGroupProps {
  /** Sets the width to "default" (for "fit-content") or "full". */
  buttonWidth?: ButtonGroupWidths;
  /** Additional className to use. */
  className?: string;
  /** ID that other components can cross reference for accessibility purposes. */
  id?: string;
  /** Set's the disabled state to all the internal `Button` components. */
  isDisabled?: boolean;
  /** Renders the layout of `Button` components in a row or column. */
  layout?: LayoutTypes;
}

const noop = () => {};

/**
 * A simple wrapper to group `Button` components together. The layout can be set
 * to row or column and the width of internal `Button` components can be set to
 * the parent's full width or the `Button`'s content width
 */
export const ButtonGroup = chakra(
  forwardRef<HTMLDivElement, React.PropsWithChildren<ButtonGroupProps>>(
    (props, ref?) => {
      const {
        buttonWidth = "default",
        children,
        className = "",
        id,
        isDisabled = false,
        layout = "row",
        ...rest
      } = props;
      const newChildren: JSX.Element[] = [];
      const { isLargerThanMobile } = useNYPLBreakpoints();
      const finalLayout = isLargerThanMobile ? layout : "column";
      const finalButtonWidth = isLargerThanMobile ? buttonWidth : "full";
      const styles = useStyleConfig("ButtonGroup", {
        buttonWidth: finalButtonWidth,
      });

      React.Children.map(
        children as JSX.Element,
        (child: React.ReactElement, key: number) => {
          if (React.isValidElement(child)) {
            if (child.type !== Button) {
              // Special case for Storybook MDX documentation.
              // @ts-ignore
              if (child.props.mdxType && child.props.mdxType === "Button") {
                noop();
              } else {
                console.warn(
                  "NYPL Reservoir ButtonGroup: Only Button components can be children of ButtonGroup."
                );
                return;
              }
            }
            const disabledProps = isDisabled ? { isDisabled } : {};
            newChildren.push(
              React.cloneElement(child, { key, ...disabledProps })
            );
          }
        }
      );

      return (
        <Stack
          className={className}
          direction={finalLayout}
          id={id}
          ref={ref}
          // Always set the spacing to "8px".
          spacing="xs"
          sx={styles}
          {...rest}
        >
          {newChildren}
        </Stack>
      );
    }
  )
);

export default ButtonGroup;
