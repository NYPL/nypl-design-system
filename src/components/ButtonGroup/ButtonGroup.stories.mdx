import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import Button from "../Button/Button";
import ButtonGroup from "./ButtonGroup";
import Heading from "../Heading/Heading";
import SimpleGrid from "../Grid/SimpleGrid";
import { getCategory } from "../../utils/componentCategories";
import DSProvider from "../../theme/provider";

<Meta
  title={getCategory("ButtonGroup")}
  component={ButtonGroup}
  decorators={[withDesign]}
  parameters={{
    jest: ["ButtonGroup.test.tsx"],
  }}
  argTypes={{
    buttonWidth: { table: { defaultValue: { summary: "default" } } },
    className: { control: false },
    id: { control: false },
    isDisabled: { table: { defaultValue: { summary: false } } },
    layout: { table: { defaultValue: { summary: "row" } } },
  }}
/>

# ButtonGroup

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.28.0`   |
| Latest            | `0.28.0`   |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Button Width](#button-width)
- [Layout](#layout)

## Overview

<Description of={ButtonGroup} />

Note: The `ButtonGroup` component does not come from Chakra because we want to
display the `Button`s in a column layout as an option, as well as control the
width.

## Component Props

<Canvas withToolbar>
  <Story
    name="ButtonGroup with Controls"
    args={{
      buttonWidth: "default",
      className: undefined,
      id: "button-id",
      isDisabled: false,
      layout: "row",
    }}
  >
    {(args) => (
      <ButtonGroup {...args}>
        <Button buttonType="secondary" id="group-1">
          Button
        </Button>
        <Button id="group-2">Submit</Button>
        <Button id="group-">Submit</Button>
        <Button id="group-4">Submit</Button>
        <Button id="group-5">Submit</Button>
      </ButtonGroup>
    )}
  </Story>
</Canvas>

<ArgsTable story="ButtonGroup with Controls" />

## Accessibility

This component is a wrapper around the `Button` component. For accessibility
information on the `Button` component, see the
[Button Accessibility](https://nypl.github.io/nypl-design-system/reservoir/v0/?path=/docs/components-form-elements-button--button-with-controls#accessibility)
documentation.

## Button Width

The `buttonWidth` prop can be used to set the width of the internal `Button`
components. The default value is `default` and it sets the width to `"fit-content"`
so the buttons will expand to its content. The `"full"` value will set the width
to the parent's width.

<Canvas>
  <DSProvider>
    <SimpleGrid columns={1}>
      <Heading level="three">default fit-content</Heading>
      <ButtonGroup>
        <Button buttonType="secondary" id="width-fit-1">
          Button
        </Button>
        <Button id="width-fit-2">Submit</Button>
      </ButtonGroup>
      <Heading level="three">full width</Heading>
      <ButtonGroup buttonWidth="full">
        <Button buttonType="secondary" id="width-full-1">
          Button
        </Button>
        <Button id="width-full-2">Submit</Button>
      </ButtonGroup>
    </SimpleGrid>
  </DSProvider>
</Canvas>

## Layout

The `layout` prop can be used to set the layout to either `row` or `column`.

<Canvas>
  <DSProvider>
    <SimpleGrid columns={1}>
      <Heading level="three">default row</Heading>
      <ButtonGroup>
        <Button buttonType="secondary" id="row-1">
          Button
        </Button>
        <Button id="row-2">Submit</Button>
      </ButtonGroup>
      <Heading level="three">column</Heading>
      <ButtonGroup layout="column">
        <Button buttonType="secondary" id="column-1">
          Button
        </Button>
        <Button id="column-2">Submit</Button>
      </ButtonGroup>
      <Heading level="three">column layout with full-width</Heading>
      <ButtonGroup buttonWidth="full" layout="column">
        <Button buttonType="secondary" id="column-full-1">
          Button
        </Button>
        <Button id="column-full-2">Submit</Button>
      </ButtonGroup>
    </SimpleGrid>
  </DSProvider>
</Canvas>
