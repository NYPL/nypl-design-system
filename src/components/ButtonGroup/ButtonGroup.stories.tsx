import { Box, VStack } from "@chakra-ui/react";
import type { Meta, StoryObj } from "@storybook/react";
import { withDesign } from "storybook-addon-designs";

import Button from "../Button/Button";
import ButtonGroup from "./ButtonGroup";
import Heading from "../Heading/Heading";

const meta: Meta<typeof ButtonGroup> = {
  title: "Components/Form Elements/ButtonGroup",
  component: ButtonGroup,
  decorators: [withDesign],
  argTypes: {
    buttonWidth: { table: { defaultValue: { summary: "default" } } },
    className: { control: false },
    id: { control: false },
    isDisabled: { table: { defaultValue: { summary: false } } },
    layout: { table: { defaultValue: { summary: "row" } } },
  },
};

export default meta;
type Story = StoryObj<typeof ButtonGroup>;

/**
 * Main Story for the ButtonGroup component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const WithControls: Story = {
  args: {
    buttonWidth: "default",
    className: undefined,
    id: "button-id",
    isDisabled: false,
    layout: "row",
  },
  parameters: {
    jest: ["ButtonGroup.test.tsx"],
  },
  render: () => (
    <ButtonGroup>
      <Button buttonType="secondary" id="group-1">
        Button
      </Button>
      <Button id="group-2">Submit</Button>
    </ButtonGroup>
  ),
};

// The following are additional ButtonGroup example Stories.
export const Width: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Box>
        <Heading level="three">default fit-content</Heading>
        <ButtonGroup>
          <Button buttonType="secondary" id="width-fit-1">
            Button
          </Button>
          <Button id="width-fit-2">Submit</Button>
        </ButtonGroup>
      </Box>
      <Box>
        <Heading level="three">full width</Heading>
        <ButtonGroup buttonWidth="full">
          <Button buttonType="secondary" id="width-full-1">
            Button
          </Button>
          <Button id="width-full-2">Submit</Button>
        </ButtonGroup>
      </Box>
    </VStack>
  ),
};

export const Layout: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Box>
        <Heading level="three">default row</Heading>
        <ButtonGroup>
          <Button buttonType="secondary" id="row-1">
            Button
          </Button>
          <Button id="row-2">Submit</Button>
        </ButtonGroup>
      </Box>
      <Box>
        <Heading level="three">column</Heading>
        <ButtonGroup layout="column">
          <Button buttonType="secondary" id="column-1">
            Button
          </Button>
          <Button id="column-2">Submit</Button>
        </ButtonGroup>
      </Box>
      <Box>
        <Heading level="three">column layout with full-width</Heading>
        <ButtonGroup buttonWidth="full" layout="column">
          <Button buttonType="secondary" id="column-full-1">
            Button
          </Button>
          <Button id="column-full-2">Submit</Button>
        </ButtonGroup>
      </Box>
    </VStack>
  ),
};

export const Disabled: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Box>
        <Heading level="three">ButtonGroup isDisabled</Heading>
        <ButtonGroup isDisabled>
          <Button buttonType="secondary" id="group-disabled-1">
            Button disabled
          </Button>
          <Button buttonType="secondary" id="group-disabled-2">
            Button disabled
          </Button>
        </ButtonGroup>
      </Box>
      <Box>
        <Heading level="three">Individual Button isDisabled</Heading>
        <ButtonGroup>
          <Button buttonType="secondary" id="individual-disabled-1">
            Button not disabled
          </Button>
          <Button buttonType="secondary" id="individual-disabled-2" isDisabled>
            Button disabled
          </Button>
        </ButtonGroup>
      </Box>
    </VStack>
  ),
};
