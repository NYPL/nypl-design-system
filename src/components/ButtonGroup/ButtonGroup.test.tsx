import * as React from "react";
import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import Button from "../Button/Button";
import ButtonGroup from "../ButtonGroup/ButtonGroup";

describe("ButtonGroup Accessibility", () => {
  it("passes axe accessibility test", async () => {
    const { container } = render(
      <ButtonGroup>
        <Button buttonType="secondary" id="btn-1">
          Button
        </Button>
        <Button id="btn-2">Submit</Button>
      </ButtonGroup>
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("ButtonGroup", () => {
  it("renders only Button components", () => {
    render(
      <ButtonGroup>
        <Button buttonType="secondary" id="btn-1">
          Button
        </Button>
        <Button id="btn-2">Submit</Button>
        <p data-test="paragraph">This is not a button</p>
      </ButtonGroup>
    );

    expect(screen.getAllByRole("button")).toHaveLength(2);
    expect(screen.queryByTestId("paragraph")).not.toBeInTheDocument();
  });

  it("sets all `Button` children as disabled if isDisabled is true", () => {
    const { rerender } = render(
      <ButtonGroup>
        <Button buttonType="secondary" id="btn-1">
          Button
        </Button>
        <Button id="btn-2">Submit</Button>
      </ButtonGroup>
    );

    expect(screen.getAllByRole("button")[0]).not.toBeDisabled();
    expect(screen.getAllByRole("button")[1]).not.toBeDisabled();

    rerender(
      <ButtonGroup isDisabled>
        <Button buttonType="secondary" id="btn-1">
          Button
        </Button>
        <Button id="btn-2">Submit</Button>
      </ButtonGroup>
    );

    expect(screen.getAllByRole("button")[0]).toBeDisabled();
    expect(screen.getAllByRole("button")[1]).toBeDisabled();
  });

  it("logs a warning when any element that is not a Button is passed ", () => {
    const warn = jest.spyOn(console, "warn");
    render(
      <ButtonGroup>
        <Button buttonType="secondary" id="btn-1">
          Button
        </Button>
        <Button id="btn-2">Submit</Button>
        <p data-test="paragraph">This is not a button</p>
      </ButtonGroup>
    );
    expect(warn).toHaveBeenCalledWith(
      "NYPL Reservoir ButtonGroup: Only Button components can be children of ButtonGroup."
    );
  });
});

describe("ButtonGroup Snapshot", () => {
  it("renders the UI snapshot correctly", () => {
    const basic = renderer
      .create(
        <ButtonGroup>
          <Button buttonType="secondary" id="btn-1">
            Button
          </Button>
          <Button id="btn-2">Submit</Button>
        </ButtonGroup>
      )
      .toJSON();
    const fullWidth = renderer
      .create(
        <ButtonGroup buttonWidth="full">
          <Button buttonType="secondary" id="btn-1">
            Button
          </Button>
          <Button id="btn-2">Submit</Button>
        </ButtonGroup>
      )
      .toJSON();
    const column = renderer
      .create(
        <ButtonGroup layout="column">
          <Button buttonType="secondary" id="btn-1">
            Button
          </Button>
          <Button id="btn-2">Submit</Button>
        </ButtonGroup>
      )
      .toJSON();
    const withChakraProps = renderer
      .create(
        <ButtonGroup p="s" color="ui.error.primary">
          <Button buttonType="secondary" id="btn-1">
            Button
          </Button>
          <Button id="btn-2">Submit</Button>
        </ButtonGroup>
      )
      .toJSON();
    const withOtherProps = renderer
      .create(
        <ButtonGroup data-testid="testid">
          <Button buttonType="secondary" id="btn-1">
            Button
          </Button>
          <Button id="btn-2">Submit</Button>
        </ButtonGroup>
      )
      .toJSON();

    expect(basic).toMatchSnapshot();
    expect(fullWidth).toMatchSnapshot();
    expect(column).toMatchSnapshot();
    expect(withChakraProps).toMatchSnapshot();
    expect(withOtherProps).toMatchSnapshot();
  });
});
