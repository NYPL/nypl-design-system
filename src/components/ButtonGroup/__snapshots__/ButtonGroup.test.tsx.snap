// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup Snapshot renders the UI snapshot correctly 1`] = `
<div
  className="chakra-stack css-6dfzd0"
>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-1"
    type="button"
  >
    Button
  </button>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-2"
    type="button"
  >
    Submit
  </button>
</div>
`;

exports[`ButtonGroup Snapshot renders the UI snapshot correctly 2`] = `
<div
  className="chakra-stack css-6dfzd0"
>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-1"
    type="button"
  >
    Button
  </button>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-2"
    type="button"
  >
    Submit
  </button>
</div>
`;

exports[`ButtonGroup Snapshot renders the UI snapshot correctly 3`] = `
<div
  className="chakra-stack css-6dfzd0"
>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-1"
    type="button"
  >
    Button
  </button>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-2"
    type="button"
  >
    Submit
  </button>
</div>
`;

exports[`ButtonGroup Snapshot renders the UI snapshot correctly 4`] = `
<div
  className="chakra-stack css-1txj82p"
>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-1"
    type="button"
  >
    Button
  </button>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-2"
    type="button"
  >
    Submit
  </button>
</div>
`;

exports[`ButtonGroup Snapshot renders the UI snapshot correctly 5`] = `
<div
  className="chakra-stack css-6dfzd0"
  data-testid="testid"
>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-1"
    type="button"
  >
    Button
  </button>
  <button
    className="chakra-button css-1xdhyk6"
    data-testid="button"
    id="btn-2"
    type="button"
  >
    Submit
  </button>
</div>
`;
